docs/hazmat/primitives/asymmetric/x25519.rst

```diff
@@ -0,0 +1,77 @@
+.. hazmat::
+
+X25519 key exchange
+===================
+
+.. currentmodule:: cryptography.hazmat.primitives.asymmetric.x25519
+
+
+X25519 is an elliptic curve `Diffie-Hellman key exchange`_ using `Curve25519`_.
+It allows two parties to jointly agree on a shared secret using an insecure
+channel.
+
+
+Exchange Algorithm
+~~~~~~~~~~~~~~~~~~
+
+For most applications the ``shared_key`` should be passed to a key
+derivation function.
+
+.. doctest::
+
+    >>> from cryptography.hazmat.backends import default_backend
+    >>> from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey
+    >>> # Generate a private key for use in the exchange.
+    >>> private_key = X25519PrivateKey.generate()
+    >>> # In a real handshake the peer_public_key will be received from the
+    >>> # other party. For this example we'll generate another private key and
+    >>> # get a public key from that. Note that in a DH handshake both peers
+    >>> # must agree on a common set of parameters.
+    >>> peer_public_key = X25519PrivateKey.generate().public_key()
+    >>> shared_key = private_key.exchange(peer_public_key)
+    >>> # For the next handshake we MUST generate another private key.
+    >>> private_key_2 = X25519PrivateKey.generate()
+    >>> peer_public_key_2 = X25519PrivateKey.generate().public_key()
+    >>> shared_key_2 = private_key_2.exchange(peer_public_key_2)
+
+Key interfaces
+~~~~~~~~~~~~~~
+
+.. class:: X25519PrivateKey
+
+    .. versionadded:: 2.0
+
+    .. classmethod:: generate()
+
+        Generate an X25519 private key.
+
+        :returns: :class:`X25519PrivateKey`
+
+    .. method:: public_key()
+
+        :returns: :class:`X25519PublicKey`
+
+    .. method:: exchange(peer_public_key)
+
+        :param X25519PublicKey peer_public_key: The public key for the
+            peer.
+
+        :returns bytes: A shared key.
+
+.. class:: X25519PublicKey
+
+    .. versionadded:: 2.0
+
+    .. classmethod:: from_public_bytes(data)
```

Comment:
Let's add an example of generating a private key, and then round tripping its public key.